Sources/SwiftCompilerPlugin/PluginMacroExpansionContext.swift

```diff
@@ -0,0 +1,253 @@
+    guard let base = self.knownSourceSyntax[node.root.id] else {
+      return nil
+    }
+    let file: String
+    switch filePathMode {
+    case .fileID: file = base.location.fileID
+    case .filePath: file = base.location.fileName
+    }
+
+    let localPosition = node.position(at: kind)
+    let localLocation = base.locationConverter.location(for: localPosition)
+
+    let positionOffset = base.location.offset
+    let lineOffset = base.location.line - 1
+    let columnOffset = localLocation.line == 1 ? base.location.column : 0
+
+    return SourceLocation(
+      line: localLocation.line! + lineOffset,
+      column: localLocation.column! + columnOffset,
+      offset: localLocation.offset + positionOffset,
+      file: file
+    )
+  }
+}
+
+fileprivate extension Syntax {
+  /// get a position in the node by `PositionInSyntaxNode`.
+  func position(at pos: PositionInSyntaxNode) -> AbsolutePosition {
+    switch pos {
+    case .beforeLeadingTrivia:
+      return self.position
+    case .afterLeadingTrivia:
+      return self.positionAfterSkippingLeadingTrivia
+    case .beforeTrailingTrivia:
+      return self.endPositionBeforeTrailingTrivia
+    case .afterTrailingTrivia:
+      return self.endPosition
+    }
+  }
+}
+
+class PluginMacroExpansionContext {
+  /// The macro expansion discriminator, which is used to form unique names
+  /// when requested.
+  ///
+  /// The expansion discriminator is combined with the `uniqueNames` counters
+  /// to produce unique names.
+  private var discriminator: String
+
+  private var sourceManger: SourceManager
+
+  /// Counter for each of the uniqued names.
+  ///
+  /// Used in conjunction with `expansionDiscriminator`.
+  private var uniqueNames: [String: Int] = [:]
+
+  /// The set of diagnostics that were emitted as part of expanding the
+  /// macro.
+  internal private(set) var diagnostics: [Diagnostic] = []
+
+  init(sourceManager: SourceManager, discriminator: String = "") {
+    self.sourceManger = sourceManager
+    self.discriminator = discriminator
+  }
+}
+
+extension PluginMacroExpansionContext: MacroExpansionContext {
+  /// Generate a unique name for use in the macro.
+  public func createUniqueName(_ providedName: String) -> TokenSyntax {
+    // If provided with an empty name, substitute in something.
+    let name = providedName.isEmpty ? "__local" : providedName
+
+    // Grab a unique index value for this name.
+    let uniqueIndex = uniqueNames[name, default: 0]
+    uniqueNames[name] = uniqueIndex + 1
+
+    // Start with the discriminator.
+    var resultString = discriminator
+
+    // Mangle the name
+    resultString += "\(name.count)\(name)"
+
+    // Mangle the operator for unique macro names.
+    resultString += "fMu"
+
+    // Mangle the index.
+    if uniqueIndex > 0 {
+      resultString += "\(uniqueIndex - 1)"
+    }
+    resultString += "_"
```

Comment:
Does this need to match any mangling logic in apple/swift? If so, I think we should add a doc comment explaining it. If not: Why do we need the `fMu` mangling prefix?

Reply:
cc: @DougGregor I copied this from `ASTGen`'s `SourceManager.MacroExpansionContext`. What's the requirement of this `createUniqueName`?